src/index.js

```diff
@@ -204,381 +291,379 @@ class Masq {
+
+            switch (json.msg) {
+              case 'authorized':
+                await this._receivedAuthorized(json)
+                break
+
+              case 'notAuthorized':
+                // if this User-app is not registered
+                await this._receiveNotAuthorized()
+                break
+
+              default:
+                // TODO change error
+                throw new MasqError(MasqError.WRONG_MESSAGE, `Unexpectedly received message with type ${json.msg}`)
+            }
+            resolve()
+          } catch (e) {
+            await this._resetLogin()
+            reject(e)
+          }
+        })
 
-      sw.on('disconnect', (peer, id) => {
-        sw.close()
-      })
+        this.loginSwDisconnectListener = async (peer, id) => {
+          debug('[masq.logIntoMasq] disconnect')
+          reject(new MasqError(MasqError.DISCONNECTED_DURING_LOGIN))
+        }
+        this.loginSw.on('disconnect', this.loginSwDisconnectListener)
 
-      sw.on('close', () => {
-        resolve()
+        this.loginSw.on('close', () => {
+          debug('[masq.logIntoMasq] close')
+        })
       })
     })
+
+    try {
+      await pr
+      this.loginSw.close()
+    } catch (e) {
+      await this._resetLogin()
+      throw e
+    }
   }
 
-  _startReplication () {
-    const discoveryKey = this.userAppDb.discoveryKey.toString('hex')
-    this.userAppRepHub = signalhub(discoveryKey, this.config.hubUrls)
-    this.userAppRepSW = this._createSwarm(this.userAppRepHub)
+  async _resetLogin () {
+    debug('[masq._resetLogin]')
+    this._deleteSessionInfo()
 
-    this.userAppRepSW.on('peer', async (peer, id) => {
-      try {
-        const stream = this.userAppDb.replicate({ live: true })
-        pump(peer, stream, peer)
-      } catch (e) {
-        if (this.isLoggedIn()) {
-          throw e
-        }
-      }
-    })
-  }
+    // clear state
+    this.stayConnected = false
+    this.loginChannel = null
+    this.loginKey = null
+    this.loginUrl = null
 
-  async _isRegistered (userId) {
-    // is registered if there is a db for this userId
-    return common.utils.dbExists(userId)
-  }
+    if (this.loginSw) {
+      // remove disconnect listener
+      if (this.loginSwDisconnectListener) {
+        this.loginSw.removeListener('disconnect', this.loginSwDisconnectListener)
+        this.loginSwDisconnectListener = null
+      }
 
-  async _requestUserAppRegister (key, peer) {
-    const msg = {
-      msg: 'registerUserApp',
-      name: this.appName,
-      description: this.appDescription,
-      imageURL: this.appImageURL
+      if (!this.loginSw.closed) {
+        await new Promise((resolve) => {
+          this.loginSw.close(() => {
+            resolve()
+          })
+        })
+      }
     }
-    let encryptedMsg = await common.crypto.encrypt(key, msg, 'base64')
-    peer.send(JSON.stringify(encryptedMsg))
+    this.loginSw = null
```

Comment:
We should move there properties to a function `init()` or something like this. This way we can call it from the constructor, and this will results in less code and far less risks of missing a property